juju/juju
# raft-leases: Handle failures writing lease state to the database

## Description of change

If the raftlease trapdoor notices that the DB leaseholder state is out of sync, it updates the database with the right holder. This is safe to do because trapdoor is only ever called when we've checked the current state in raft is good.

To avoid load on the DB for something that should only happen occasionally, we only check the leaseholder state on the second attempt at building+running a leadership transaction. This requires that `lease.Token.Check`, `lease.Trapdoor` and `leadership.Token.Check` get passed in the attempt (we can't keep it as closure state because the trapdoor is recreated each time).

The lease manager manifold now needs the state worker, because the trapdoor factory needs a *state.State so that it can update leaseholders when needed.

## QA steps

* Compile this with the wrench in notifyTarget.Claimed uncommented. 
* Bootstrap a controller and enable HA.
* Run `echo fail-claim > /var/lib/juju/wrench/raftlease-notifytarget` on each controller to prevent the DB from being updated when a lease is claimed in raft.
* Watch the controller debug-log for errors.
* Deploy leadership-flex (https://github.com/mitechie/leadership-charm) - it sets leadership settings when a unit is elected leader.
* When the first unit is deployed you should see the wrench error in the controller log (on the raft leader) indicating that we couldn't update the database with the new lease.
* The new unit should still start up successfully, including setting leadership settings - its state will be "ready to go". No hooks should fail, and status should indicate it's the leader.
* The database should contain the leaseholder record for leadership-flex.
* Add another unit of leadership-flex.
* Once it's running, remove the original unit.
* When the lease expires the new unit should claim the leadership, no hooks fail.
* The controller error log will show the claim couldn't be written to the db when the raft change occurred (because of the wrench).
* Status will indicate the new unit is the leader, and the database will indicate the new lease holder.

## Documentation changes

None

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1810331
